frontend/src/responseQuizView/responseQuizView.js

```diff
@@ -0,0 +1,343 @@
+const ErrorDescriptions = {
+    'backwards': 'The letters are reversed',
+    'verb-conjugation': 'One or more verbs are incorrectly conjugated',
+    'comma-splice': 'Two independent clauses are incorrectly joined by a comma',
+    'run-on': 'There is a run on sentence',
+    'homophone': 'Misuse of a similar sounding word (homophone)',
+};
+
+export class ResponseQuizView extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            data: null,
+            question: 1,
+            userAnswers: {},
+            score: 0,
+            graded: false,
+        };
+    }
+
+    async componentDidMount() {
+        try {
+            const response = await fetch(`/api/get_response_quiz_data/${this.props.textID}/`);
+            const data = await response.json();
+            this.setState({ data: data });
+        } catch (e) {
+            console.log(e);
+        }
+    }
+
+    nextQuestion() {
+        const nextQuestionNumber = this.state.question + 1;
+        if (nextQuestionNumber <= this.state.data.length) {
+            this.setState({
+                question: nextQuestionNumber,
+            });
+        } else {
+            console.log();
+        }
+    }
+
+    prevQuestion() {
+        const prevQuestionNumber = this.state.question - 1;
+        if (prevQuestionNumber > 0) {
+            this.setState({
+                question: prevQuestionNumber,
+            });
+        } else {
+            console.log();
+        }
+    }
+
+    getUnanswered() {
+        let numUnanswered = 0;
+        const answers = this.state.userAnswers;
+        for (let i = 1; i <= this.state.data.length; i++) {
+            if (answers[i] === undefined) {
+                numUnanswered++;
+            }
+        }
+        return numUnanswered;
+    }
+
+    gradeQuiz() {
+        let cSubmit;
+        if (this.getUnanswered() === 0) {
+            cSubmit = window.confirm('Are you sure that you want to submit your quiz? Your'
+                + ' answers are final.');
+        } else {
+            cSubmit = window.confirm('Are you sure that you want to submit your quiz? You'
+                + ' have ' + this.getUnanswered() + ' unanswered questions(s).');
+        }
+        if (cSubmit === true) {
+            this.setState({ graded: true });
+            let score = 0;
+            const answers = this.state.userAnswers;
+            for (let i = 0; i < this.state.data.length; i++) {
+                if (answers[i + 1] === this.state.data[i].answer) {
+                    score += 1;
+                }
+            }
+            this.setState({ score: score });
+        } else {
+            console.log();
+        }
+
+        // Brings the user back to the first question (beginning of the quiz) upon submission.
+        this.setState({ question: 1 });
+    }
+
+    onProgressBarClick = (event) => {
+        this.setState({
+            ...this.state,
```

Comment:
What's the purpose of spreading `this.state` here?

Reply:
Removed.